File: TheaterSuggestions/CSharp/SeatsSuggestions/Seat.cs
using System;
using System.Collections.Generic;
using Value;

namespace SeatsSuggestions
{
    public class Seat : ValueType<Seat>
    {
        public Seat(string rowName, uint number, PricingCategory pricingCategory, SeatAvailability seatAvailability)
        {
            RowName = rowName;
            Number = number;
            PricingCategory = pricingCategory;
            SeatAvailability = seatAvailability;
        }

        public string RowName { get; }
        public uint Number { get; }
        public PricingCategory PricingCategory { get; }
        public SeatAvailability SeatAvailability { get; }

        public bool IsAvailable()
        {
            return SeatAvailability == SeatAvailability.Available;
        }

        public override string ToString()
        {
            return $"{RowName}{Number}";
        }

        public bool MatchCategory(PricingCategory pricingCategory)
        {
            if (pricingCategory == PricingCategory.Mixed)
            {
                return true;
            }

            return PricingCategory == pricingCategory;
        }

        public Seat Allocate()
        {
            if (SeatAvailability == SeatAvailability.Available)
            {
                return new Seat(RowName, Number, PricingCategory, SeatAvailability.Allocated);
            }

            return this;
        }

        public bool SameSeatLocation(Seat seat)
        {
            return RowName == seat.RowName && Number == seat.Number;
        }

        protected override IEnumerable<object> GetAllAttributesToBeUsedForEquality()
        {
            return new object[] {RowName, Number, PricingCategory, SeatAvailability};
        }

        public bool IsAdjacentWith(uint number)
        {
            return Number + 1 == number || Number - 1 == number;
        }

        public int ComputeDistanceFromRowCentroid(int rowSize)
        {
            var distance = 0;
            var centroidIndex = Math.Abs(rowSize / 2);

            if (rowSize % 2 == 0)
            {
                if (Number == centroidIndex || Number == centroidIndex + 1) return 0;
                if (Number < centroidIndex)
                {
                    distance = (int) Math.Abs(Number - centroidIndex);
                }
                else
                {
                    distance = (int) Math.Abs(Number - centroidIndex) - 1;
                }
            }
            else
            {
                distance = (int) Math.Abs(Number - centroidIndex);
            }

            return distance;
        }
    }
}

File: TheaterSuggestions/Java/SeatsSuggestions/src/main/java/com/baasie/SeatsSuggestions/Seat.java
package com.baasie.SeatsSuggestions;

import lombok.EqualsAndHashCode;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;

@EqualsAndHashCode
public class Seat {

    private final String rowName;
    private final int number;
    private final PricingCategory pricingCategory;
    private final SeatAvailability seatAvailability;

    public Seat(String rowName, int number, PricingCategory pricingCategory, SeatAvailability seatAvailability) {
        this.rowName = rowName;
        this.number = number;
        this.pricingCategory = pricingCategory;
        this.seatAvailability = seatAvailability;
    }

    public boolean isAvailable() {
        return seatAvailability == SeatAvailability.Available;
    }

    public boolean matchCategory(PricingCategory pricingCategory) {
        if (pricingCategory == PricingCategory.Mixed) {
            return true;
        }

        return this.pricingCategory == pricingCategory;
    }

    Seat allocate() {
        if (seatAvailability == SeatAvailability.Available) {
            return new Seat(rowName, number, pricingCategory, SeatAvailability.Allocated);
        }
        return this;
    }

    public boolean sameSeatLocation(Seat seat) {
        return rowName.equals(seat.rowName) && number == seat.number;
    }

    boolean isAdjacentWith(List<Seat> seats) {

        List<Seat> orderedSeats = seats.stream()
                .sorted(Comparator.comparing(Seat::number))
                .collect(Collectors.toCollection(ArrayList::new));

        for (Seat seat : orderedSeats) {
            if (number + 1 == seat.number || number - 1 == seat.number)
                return true;
        }
        return false;
    }

    public SeatAvailability seatAvailability() {
        return seatAvailability;
    }

    public String rowName() {
        return rowName;
    }

    public int number() {
        return number;
    }

    public PricingCategory pricingCategory() {
        return pricingCategory;
    }

    @Override
    public String toString() {
        return rowName + number;
    }
}